types/time.js

```diff
@@ -18,12 +18,64 @@ class TimeType extends Commando.ArgumentType {
 			raid_creation_time = raid_exists ?
 				moment(Raid.getRaid(message.channel.id).creation_time) :
 				now,
+			incubation_duration = is_ex_raid ?
+				settings.exclusive_raid_incubate_duration :
+				settings.standard_raid_incubate_duration,
 			hatched_duration = is_ex_raid ?
 				settings.exclusive_raid_hatched_duration :
-				settings.standard_raid_hatched_duration,
-			last_possible_time = raid_creation_time.clone().add(is_ex_raid ?
-				settings.exclusive_raid_duration :
-				settings.default_raid_duration, 'minutes');
+				settings.standard_raid_hatched_duration;
+
+		let first_possible_time,
+			max_duration,
+			last_possible_time;
+
+		// Figure out valid first and last possible times for this time
+		switch (arg.key) {
+			case TimeParameter.START:
+				// Start time - valid range is now (or hatch time if it exists, whichever is later)
+				// through raid's end time
+				const raid = Raid.getRaid(message.channel.id),
+					hatch_time = raid ?
+						raid.hatch_time :
+						undefined,
+					end_time = raid ?
+						raid.end_time :
+						undefined;
+
+				if (hatch_time) {
+					const hatch_time_moment = moment(hatch_time);
+
+					first_possible_time = now.isAfter(hatch_time_moment) ?
+						now :
+						hatch_time_moment;
+				} else {
+					first_possible_time = now;
+				}
+
+				const raid_end_time = end_time ?
+					moment(end_time) :
+					raid_creation_time.clone().add(incubation_duration + hatched_duration, 'minutes');
+
+				max_duration = incubation_duration + hatched_duration;
+				last_possible_time = raid_end_time;
+				break;
+
+			case TimeParameter.HATCH: {
+				// Hatch time - valid range is up to hatched duration in the past
+				// through incubation period past raid creation time
+				first_possible_time = now.clone().add(-hatched_duration, 'minutes');
+				max_duration = incubation_duration;
+				last_possible_time = raid_creation_time.clone().add(max_duration, 'minutes');
+				break;
+			}
+
+			case TimeParameter.END:
+				// End time - valid range is now through incubation plus hatch duration past creation time
+				first_possible_time = now;
+				max_duration = incubation_duration + hatched_duration;
```

Comment:
so the maximum end time / duration for a raid is now 2 hours? just making sure this is on purpose

Reply:
It always was, this just makes how that duration comes about more explicit, clearer, and more configurable if Niantic changes things down the line. :)